delphi-frontend/src/main/java/au/com/integradev/delphi/cfg/api/Block.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Sonar Delphi Plugin
+ * Copyright (C) 2024 Integrated Application Development
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package au.com.integradev.delphi.cfg.api;
+
+import java.util.List;
+import java.util.Set;
+import org.sonar.plugins.communitydelphi.api.ast.DelphiNode;
+
+public interface Block {
+  /** Unique block identifier */
+  int getId();
+
+  /** Successor blocks of the block */
+  Set<Block> getSuccessorBlocks();
+
+  /** Predecessors of the block */
+  Set<Block> getPredecessorBlocks();
+
+  /** Successors handler of the block */
+  Successors getSuccessors();
```

Comment:
I'm feeling a bit weird about:
* `getSuccessorBlocks` and `getSuccessors` coexisting at this level of the API
* the asymmetry due to no neatly-typed `getPredecessors` matching up with `getSuccessors`

Reply:
You commented on this above, so I'm just getting some initial thoughts out about the current API shape.

At minimum, maybe you can speak to the reason for the asymmetry here?
`Successors` got its own `Successors` type hierarchy, while predecessors are just the basic `Set<Block>`.

Reply:
Thinking about it a little more, I think the `Successors` API is an indirect way of being able to check what type of block a block is. All the while maintaining the object permanence as mentioned.
The `getSuccessorBlocks` endpoint is the opaque way of accessing the successors and is what would be primarily utilised to traverse the graph. The `getSuccessors` on the other hand is a more strongly typed idea, indicated what kinds of successors there are.

Reasoning from my use of it here, and the usages of it in `sonar-java`, the predecessors are just a basic set because the graph is directed. A node only really knows about where it goes next, the fact the predecessors are there at all is a matter of convenience, I would suggest.

I think a solution that makes a lot of sense would be one that incorporates a few things I mentioned:
* Move the type hierarchy to the `Block` type
* Use IDs in the construction, and only construct the object referenced links after pruning
* Alter the API back to the original `sonar-java` approach of just `getSuccessors` and `getPredecessors` on the main `Block`

Let me know what you think.

Reply:
Yeah, I'm on board with this idea.
It takes the good part of the SonarJava API while also avoiding boolean flag hell on `Block`.